README.md
# pyramid_webpack
Pyramid extension for managing assets with Webpack


CHANGES.rst
Changelog
=========

0.1.2 - 2017/9/3
----------------
* Fix `#4 <https://github.com/stevearc/pyramid_webpack/issues/4>`__: Issue with Jinja2 rendering
* Created a `cookiecutter <https://cookiecutter.readthedocs.io/en/latest/>`__ to make it easier to get started: https://github.com/stevearc/pyramid-cookiecutter-webpack

0.1.1 - 2016/10/27
------------------
* Add cache_max_age config option

0.1.0 - 2016/10/23
------------------
* Initial release


tests/__init__.py
""" Tests for pyramid_webpack """
import os
import inspect
import re

import json
import shutil
import tempfile
import webtest
from mock import MagicMock
from pyramid.config import Configurator
from pyramid.renderers import render_to_response
from six.moves.queue import Queue, Empty  # pylint: disable=E0401
from threading import Thread

from pyramid_webpack import WebpackState, Webpack, StaticResource


try:
    import unittest2 as unittest  # pylint: disable=E0401
except ImportError:
    import unittest


def load_stats(state, queue, *args, **kwargs):
    """ Load stats and put it into a queue """
    # Put a sentinel in the queue when the thread starts
    queue.put(object())
    stats = state.load_stats(*args, **kwargs)
    queue.put(stats)


def run_load_stats(state, *args, **kwargs):
    """ Run load_stats() in a thread """
    queue = Queue()
    thread_args = [state, queue]
    thread_args.extend(args)
    thread = Thread(target=load_stats, args=thread_args, kwargs=kwargs)
    thread.daemon = True
    thread.start()
    # Wait until the thread is definitely started
    queue.get(True)
    return queue


class TempDirTest(unittest.TestCase):

    """ Test class that provides filesystem helpers """

    def setUp(self):
        super(TempDirTest, self).setUp()
        self._tempdir = tempfile.mkdtemp()

    def tearDown(self):
        super(TempDirTest, self).tearDown()
        shutil.rmtree(self._tempdir)

    def _write(self, filename, data):
        """ Write json data to a file """
        fullpath = os.path.join(self._tempdir, filename)
        with open(fullpath, 'w') as ofile:
            json.dump(data, ofile)
        return fullpath


class TestWebpackState(TempDirTest):
    """ Tests for the WebpackState """

    def test_load_stats(self):
        """ State loads stats from a json file """
        data = {'a': 'b'}
        stats_file = self._write('stats.json', data)
        settings = {
            'webpack.stats_file': stats_file,
        }
        state = WebpackState(settings)
        stats = state.load_stats()
        self.assertEqual(stats, data)

    def test_load_stats_as_asset(self):
        """
        State loads stats from a json file specified in a:b/c.json notation

        Tests the regression where loading the stats file specified in the
        notation didn't work for Python 3 (before 3.6).
        """
        settings = {
            'webpack.stats_file': 'tests:test-stats.json',
        }
        state = WebpackState(settings)
        stats = state.load_stats()
        self.assertEqual(stats, {'a': 'b'})

    def test_missing_stats(self):
        """ raise IOError if stats file is missing """
        state = WebpackState({})
        with self.assertRaises(IOError):
            state.load_stats()

    def test_cache_stats(self):
        """ When cache=True, cache the stats file """
        data = {'a': 'b'}
        stats_file = self._write('stats.json', data)
        settings = {
            'webpack.stats_file': stats_file,
        }
        state = WebpackState(settings)
        stats = state.load_stats(cache=True)
        self.assertEqual(data, stats)
        with open(stats_file, 'w') as ofile:
            json.dump({'b': 'c'}, ofile)
        second_stats = state.load_stats(cache=True)
        self.assertEqual(second_stats, stats)

    def test_no_cache_stats(self):
        """ When cache=False, don't cache the stats file """
        stats_file = self._write('stats.json', {'a': 'b'})
        settings = {
            'webpack.stats_file': stats_file,
        }
        state = WebpackState(settings)
        state.load_stats(cache=False)
        data = {'b': 'c'}
        with open(stats_file, 'w') as ofile:
            json.dump(data, ofile)
        stats = state.load_stats(cache=False)
        self.assertEqual(stats, data)

    def test_multiple_configs(self):
        """ Multiple webpack states can have their own configs """
        settings = {
            'webpack.stats_file': 'foo',
            'webpack.other.stats_file': 'bar',
        }
        state = WebpackState(settings)
        self.assertTrue(state.stats_file.path.endswith(':foo'))
        other_state = WebpackState(settings, name='other')
        self.assertTrue(other_state.stats_file.path.endswith(':bar'))

    def test_no_wait_for_compile(self):
        """ The load_stats() call doesn't block if wait=False """
        data = {'status': 'compiling'}
        stats_file = self._write('stats.json', data)
        settings = {
            'webpack.stats_file': stats_file,
        }
        state = WebpackState(settings)
        queue = run_load_stats(state, wait=False)
        # Blocks & doesn't raise an exception
        stats = queue.get(True, 0.1)
        self.assertEqual(stats, data)

    def test_wait_for_compile(self):
        """ The load_stats() call blocks until webpack is done compiling """
        stats_file = self._write('stats.json', {'status': 'compiling'})
        settings = {
            'webpack.stats_file': stats_file,
        }
        state = WebpackState(settings)
        queue = run_load_stats(state, wait=True)
        with self.assertRaises(Empty):
            queue.get(True, 0.1)

        stats = {'status': 'done'}
        with open(stats_file, 'w') as ofile:
            json.dump(stats, ofile)
        data = queue.get(True, 5)
        self.assertEqual(data, stats)

    def test_compile_timeout(self):
        """ The load_stats() call will timeout if compile takes too long """
        stats_file = self._write('stats.json', {'status': 'compiling'})
        settings = {
            'webpack.stats_file': stats_file,
            'webpack.timeout': 0.5,
        }
        state = WebpackState(settings)
        with self.assertRaises(RuntimeError):
            state.load_stats(wait=True)

    def test_load_bad_data(self):
        """ load_stats() raises ValueError if json data is bad """
        stats_file = self._write('stats.json', {})
        with open(stats_file, 'a') as ofile:
            ofile.write('aaaaa')
        settings = {
            'webpack.stats_file': stats_file,
        }
        state = WebpackState(settings)
        with self.assertRaises(ValueError):
            state.load_stats()

    def test_abs_static_view(self):
        """ Absolute bundle directory paths are stored unmodified """
        settings = {
            'webpack.bundle_dir': '/foo/bar/baz',
        }
        state = WebpackState(settings)
        self.assertEqual(state.static_view_path, '/foo/bar/baz')

    def test_unspecified_relative_bundle(self):
        """ Relative bundle_dir paths with no package are given one """
        settings = {
            'webpack.bundle_dir': 'bundles',
        }
        state = WebpackState(settings, 'mypackage')
        self.assertEqual(state.static_view_path, 'mypackage:bundles')

    def test_static_package_resource(self):
        """ StaticResource can load a package resource """
        resource = StaticResource('pyramid_webpack:jinja2ext.py')
        import pyramid_webpack.jinja2ext
        with resource.open() as i:
            self.assertEqual(i.read(),
                             inspect.getsource(pyramid_webpack.jinja2ext))

    def test_future_expire(self):
        """ cache_max_age = future uses 10 year expiration """
        settings = {
            'webpack.cache_max_age': 'future',
        }
        state = WebpackState(settings, 'mypackage')
        self.assertTrue(state.cache_max_age > 100000)

    def test_custom_expire(self):
        """ cache_max_age can specify view expiration """
        settings = {
            'webpack.cache_max_age': '1234',
        }
        state = WebpackState(settings, 'mypackage')
        self.assertEqual(state.cache_max_age, 1234)

    def test_default_expire_debug(self):
        """ cache_max_age defaults to None in debug mode """
        settings = {
            'webpack.debug': 'true',
        }
        state = WebpackState(settings, 'mypackage')
        self.assertIsNone(state.cache_max_age)

    def test_default_expire(self):
        """ cache_max_age defaults to 3600 in non-debug mode """
        settings = {
            'webpack.debug': 'false',
        }
        state = WebpackState(settings, 'mypackage')
        self.assertEqual(state.cache_max_age, 3600)


class TestWebpack(unittest.TestCase):

    """ Test class for the Webpack functions """

    def setUp(self):
        super(TestWebpack, self).setUp()
        self.request = MagicMock()
        self.webpack = Webpack(self.request)
        self.webpack.state = WebpackState({})
        self.stats = {
            'status': 'done',
            'chunks': {
                'main': [
                    {
                        'name': 'main.js',
                        'path': '/static/main.js',
                    }
                ],
            },
        }
        self.webpack.state._load_stats = MagicMock()
        self.webpack.state._load_stats.return_value = self.stats

    def test_get_bundle(self):
        """ get_bundle() returns the chunks with a 'url' key added """
        bundle = self.webpack.get_bundle('main')
        self.assertEqual(bundle, self.stats['chunks']['main'])

    def test_filter_extensions(self):
        """ get_bundle() can filter by file extension """
        chunk = {
            'name': 'main.css',
            'path': '/static/main.css',
        }
        self.stats['chunks']['main'].append(chunk)
        bundle = self.webpack.get_bundle('main', '.css')
        self.assertEqual(bundle, [chunk])

    def test_filter_multiple_extensions(self):
        """ get_bundle() can filter by multiple file extensions """
        chunk = {
            'name': 'main.css',
            'path': '/static/main.css',
        }
        self.stats['chunks']['main'].append(chunk)
        bundle = self.webpack.get_bundle('main', '.js .css')
        self.assertEqual(bundle, self.stats['chunks']['main'])

    def test_filter_ignore(self):
        """ get_bundle() can ignore files by glob """
        chunk = {
            'name': 'main.css',
            'path': '/static/main.css',
        }
        self.stats['chunks']['main'].append(chunk)
        self.webpack.state.ignore = ['*.css']
        bundle = self.webpack.get_bundle('main')
        self.assertEqual(bundle, self.stats['chunks']['main'][:1])

    def test_filter_ignore_re(self):
        """ get_bundle() can ignore files by regular expression """
        chunk = {
            'name': 'main.css',
            'path': '/static/main.css',
        }
        self.stats['chunks']['main'].append(chunk)
        self.webpack.state.ignore_re = [re.compile(r'.*\.css')]
        bundle = self.webpack.get_bundle('main')
        self.assertEqual(bundle, self.stats['chunks']['main'][:1])

    def test_public_path(self):
        """ pulicPath in a chunk becomes the url """
        url = 'https://assets.cdn.com/main.js'
        self.stats['chunks']['main'][0]['publicPath'] = url
        bundle = self.webpack.get_bundle('main')
        self.assertEqual(bundle[0]['url'], url)

    def test_bad_bundle(self):
        """ Getting a nonexistant bundle raises an exception """
        with self.assertRaises(KeyError):
            self.webpack.get_bundle('nope')

    def test_error_status(self):
        """ If stats has error status, raise an error """
        self.stats['status'] = 'error'
        self.stats['error'] = 'FrobnicationError'
        self.stats['message'] = 'Retro quarks emitted during Frobnication'
        with self.assertRaises(RuntimeError):
            self.webpack.get_bundle('main')

    def test_bad_status(self):
        """ If stats has unknown status, raise an error """
        self.stats['status'] = 'wat'
        with self.assertRaises(RuntimeError):
            self.webpack.get_bundle('main')

    def test_missing_state(self):
        """ Raise an error if no WebpackState found """
        req = MagicMock()
        req.registry.webpack = {}
        with self.assertRaises(RuntimeError):
            Webpack(req)


def _get_bundle(request):
    """ Route view for the test webapp """
    config = request.matchdict['config']
    bundle_name = request.matchdict['bundle']
    bundle = request.webpack(config).get_bundle(bundle_name)
    renderer = request.params.get('renderer')
    if renderer:
        return render_to_response(renderer, {})
    else:
        return bundle


class TestWebapp(TempDirTest):
    """ Pyramid app tests """

    def setUp(self):
        super(TestWebapp, self).setUp()
        self.stats1 = {
            'status': 'done',
            'chunks': {
                'main': [
                    {
                        'name': 'main.js',
                        'path': '/static/main.js',
                    },
                ],
            },
        }
        self.stats2 = {
            'status': 'done',
            'chunks': {
                'libs': [
                    {
                        'name': 'libs.js',
                        'path': '/static/libs.js',
                    },
                ],
            },
        }

        settings = {
            'pyramid.includes': ['pyramid_jinja2', 'pyramid_webpack'],
            'jinja2.extensions': ['pyramid_webpack.jinja2ext:WebpackExtension'],
            'jinja2.directories': ['tests:templates/'],
            'webpack.debug': True,
            'webpack.stats_file': self._write('stats1.json', self.stats1),
            'webpack.configs': ['other'],
            'webpack.other.stats_file': self._write('stats2.json', self.stats2),
        }
        config = Configurator(settings=settings)

        config.add_route('bundle', '/bundle/{config}/{bundle}')
        config.add_view(_get_bundle, route_name='bundle', renderer='json')

        app = config.make_wsgi_app()
        self.app = webtest.TestApp(app)

    def tearDown(self):
        self.app.reset()

    def test_get_bundle(self):
        """ get_bundle() returns a list of all chunks in the bundle """
        res = self.app.get('/bundle/DEFAULT/main')
        bundle = json.loads(res.body.decode('utf-8'))
        expected = self.stats1['chunks']['main'][0]
        self.assertEqual(len(bundle), 1)
        self.assertEqual(bundle[0]['name'], expected['name'])
        self.assertEqual(bundle[0]['path'], expected['path'])
        self.assertTrue('url' in bundle[0])

    def test_get_second_bundle(self):
        """ get_bundle() works with the secondary webpack configs """
        res = self.app.get('/bundle/other/libs')
        bundle = json.loads(res.body.decode('utf-8'))
        expected = self.stats2['chunks']['libs'][0]
        self.assertEqual(len(bundle), 1)
        self.assertEqual(bundle[0]['name'], expected['name'])
        self.assertEqual(bundle[0]['path'], expected['path'])
        self.assertTrue('url' in bundle[0])

    def test_jinja2(self):
        """ The jinja2 extension can use 'webasset' blocks """
        res = self.app.get('/bundle/DEFAULT/main?renderer=paths.jinja2')
        expected = self.stats1['chunks']['main'][0]
        self.assertEqual(res.body.decode('utf-8'), expected['path'] + '\n')

    def test_jinja2_ext(self):
        """ The jinja2 extension can specify file extensions """
        res = self.app.get('/bundle/other/libs?renderer=paths2.jinja2')
        expected = self.stats2['chunks']['libs'][0]
        self.assertEqual(res.body.decode('utf-8'), expected['path'] + '\n')


pyramid_webpack/jinja2ext.py
""" Jinja2 extension for pyramid_webpack """
from __future__ import unicode_literals

import six
from pyramid.threadlocal import get_current_request

from jinja2 import nodes
from jinja2.ext import Extension


class WebpackExtension(Extension):

    """
    Extension for jinja2.

    Examples
    --------
    ::

        {% webpack 'main' %}
            <link rel="stylesheet" type="text/css" href="{{ ASSET.url }}">
        {% endwebpack %}

    ::

        {% webpack 'other_config:main', 'js', 'js.gz' %}
            <script type="text/javascript" src="{{ ASSET.url }}"></script>
        {% endwebpack %}

    """

    tags = set(['webpack'])

    def parse(self, parser):
        # the first token is the token that started the tag.  In our case
        # we only listen to ``'webpack'`` so this will be a name token with
        # `webpack` as value.  We get the line number so that we can give
        # that line number to the nodes we create by hand.
        lineno = six.next(parser.stream).lineno
        ctx_ref = nodes.ContextReference()

        # Parse a single expression that is the 'bundle' or 'config:bundle'
        args = [ctx_ref, parser.parse_expression()]

        # if there is a comma, the user provided an 'extensions' arg
        if parser.stream.skip_if('comma'):
            args.append(parser.parse_expression())
        else:
            args.append(nodes.Const(None))

        # now we parse the body of the cache block up to `endwebpack` and
        # drop the needle (which would always be `endwebpack` in that case)
        body = parser.parse_statements(['name:endwebpack'], drop_needle=True)

        call_args = [nodes.Name('ASSET', 'param')]

        return nodes.CallBlock(self.call_method('_get_graph', args), call_args,
                               [], body).set_lineno(lineno)

    def _get_graph(self, ctx, bundle, extensions, caller=None):
        """ Run a graph and render the tag contents for each output """
        request = ctx.get('request')
        if request is None:
            request = get_current_request()
        if ':' in bundle:
            config_name, bundle = bundle.split(':')
        else:
            config_name = 'DEFAULT'
        webpack = request.webpack(config_name)
        assets = (caller(a) for a in webpack.get_bundle(bundle, extensions))
        return ''.join(assets)


pyramid_webpack/__init__.py
""" pyramid_webpack """
import fnmatch
import posixpath
import re
import time
from io import StringIO

import json
import six
from pkg_resources import resource_string
from pyramid.decorator import reify
from pyramid.settings import asbool, aslist


__version__ = '0.1.3'


SENTINEL = object()


@six.python_2_unicode_compatible
class StaticResource(object):
    """ Wrapper around a filepath or asset path """

    def __init__(self, path):
        self.path = path

    @classmethod
    def create(cls, path, root_package):
        """ Create a StaticResource, setting the package if needed """
        if ':' not in path and not path.startswith('/'):
            return cls("{0}:{1}".format(root_package, path))
        return cls(path)

    def open(self):
        """ Open a stream object to the resource data """
        if self.path.startswith('/'):
            # Absolute path
            filepath = self.path
            return open(filepath, 'r')
        else:
            # Asset specification
            package, filename = self.path.split(':')
            contents = resource_string(package, filename)
            return StringIO(contents.decode('utf-8'))

    def __str__(self):
        return "Resource('{0}')".format(self.path)


class WebpackState(object):

    """ Wrapper for all webpack configuration and cached data """

    def __init__(self, settings, root_package_name=__package__, name='DEFAULT'):
        self.name = name
        self._settings = settings
        self._stats = None
        self.debug = asbool(self._get_setting('debug', False))
        self.static_view = asbool(self._get_setting('static_view', True,
                                                    inherit=False))
        bundle_dir = self._get_setting('bundle_dir', None, inherit=False)
        if bundle_dir is None:
            self.static_view_path = 'webpack-{0}'.format(name)
        elif ':' not in bundle_dir and not bundle_dir.startswith('/'):
            asset_spec = '{0}:{1}'.format(root_package_name, bundle_dir)
            self.static_view_path = asset_spec
        else:
            self.static_view_path = bundle_dir
        self.static_view_name = self._get_setting('static_view_name',
                                                  'webpack-{0}'.format(name),
                                                  inherit=False)
        stats_file_path = self._get_setting('stats_file', 'webpack-stats.json',
                                            inherit=False)
        self.stats_file = StaticResource.create(stats_file_path,
                                                root_package_name)
        self.timeout = float(self._get_setting('timeout', 0))
        max_age = self._get_setting('cache_max_age', None)
        if max_age is None:
            if not self.debug:
                max_age = 3600
        elif max_age == 'future':
            # 10 years in the future
            max_age = 10 * 365 * 24 * 60 * 60
        else:
            max_age = float(max_age)
        self.cache_max_age = max_age
        self.ignore = aslist(self._get_setting('ignore',
                                               [r'*.hot-update.js', r'*.map']))
        ignore_re = aslist(self._get_setting('ignore_re', []))
        self.ignore_re = [re.compile(p) for p in ignore_re]

    def _get_setting(self, setting, default=None, name=None, inherit=True):
        """ Helper function to fetch settings, inheriting from the base """
        if name is None:
            name = self.name
        if name == 'DEFAULT':
            return self._settings.get('webpack.{0}'.format(setting), default)
        else:
            val = self._settings.get('webpack.{0}.{1}'.format(name, setting),
                                     SENTINEL)
            if val is SENTINEL:
                if inherit:
                    return self._get_setting(setting, default, 'DEFAULT')
                else:
                    return default
            else:
                return val

    def load_stats(self, cache=None, wait=None):
        """ Load and cache the webpack-stats file """
        if cache is None:
            cache = not self.debug
        if wait is None:
            wait = self.debug
        if not cache or self._stats is None:
            self._stats = self._load_stats()
            start = time.time()
            while wait and self._stats.get('status') == 'compiling':
                if self.timeout and (time.time() - start > self.timeout):
                    raise RuntimeError("Webpack {0!r} timed out while compiling"
                                       .format(self.stats_file.path))
                time.sleep(0.1)
                self._stats = self._load_stats()
        return self._stats

    def _load_stats(self):
        """ Load the webpack-stats file """
        for attempt in range(0, 3):
            try:
                with self.stats_file.open() as f:
                    return json.load(f)
            except ValueError:
                # If we failed to parse the JSON, it's possible that the
                # webpack process is writing to it concurrently and it's in a
                # bad state. Sleep and retry.
                if attempt < 2:
                    time.sleep(attempt * 0.2)
                else:
                    raise
            except IOError:
                raise IOError(
                    "Could not read stats file {0}. Make sure you are using the "
                    "webpack-bundle-tracker plugin" .format(self.stats_file))


class Webpack(object):

    """ Wrapper object for the public webpack API """

    def __init__(self, request, name='DEFAULT'):
        self.name = name
        self._request = request
        self.state = request.registry.webpack.get(name)
        if self.state is None:
            raise RuntimeError("Unknown webpack config {0!r}".format(name))

    @reify
    def stats(self):
        """ Load and cache the webpack stats file """
        return self.state.load_stats()

    def _chunk_filter(self, extensions):
        """ Create a filter from the extensions and ignore files """
        if isinstance(extensions, six.string_types):
            extensions = extensions.split()

        def _filter(chunk):
            """ Exclusion filter """
            name = chunk['name']
            if extensions is not None:
                if not any(name.endswith(e) for e in extensions):
                    return False
            for pattern in self.state.ignore_re:
                if pattern.match(name):
                    return False
            for pattern in self.state.ignore:
                if fnmatch.fnmatchcase(name, pattern):
                    return False
            return True
        return _filter

    def _add_url(self, chunk):
        """ Add a 'url' property to a chunk and return it """
        if 'url' in chunk:
            return chunk
        public_path = chunk.get('publicPath')
        if public_path:
            chunk['url'] = public_path
        else:
            fullpath = posixpath.join(self.state.static_view_path,
                                      chunk['name'])
            chunk['url'] = self._request.static_url(fullpath)
        return chunk

    def get_bundle(self, bundle_name, extensions=None):
        """ Get all the chunks contained in a bundle """
        if self.stats.get('status') == 'done':
            bundle = self.stats.get('chunks', {}).get(bundle_name, None)
            if bundle is None:
                raise KeyError('No such bundle {0!r}.'.format(bundle_name))
            test = self._chunk_filter(extensions)
            return [self._add_url(c) for c in bundle if test(c)]
        elif self.stats.get('status') == 'error':
            raise RuntimeError("{error}: {message}".format(**self.stats))
        else:
            raise RuntimeError(
                "Bad webpack stats file {0} status: {1!r}"
                .format(self.state.stats_file, self.stats.get('status')))


def get_webpack(request, name='DEFAULT'):
    """
    Get the Webpack object for a given webpack config.

    Called at most once per request per config name.
    """
    if not hasattr(request, '_webpack_map'):
        request._webpack_map = {}
    wp = request._webpack_map.get(name)
    if wp is None:
        wp = request._webpack_map[name] = Webpack(request, name)
    return wp


def includeme(config):
    """ Add pyramid_webpack methods and config to the app """
    settings = config.registry.settings
    root_package_name = config.root_package.__name__
    config.registry.webpack = {
        'DEFAULT': WebpackState(settings, root_package_name)
    }
    for extra_config in aslist(settings.get('webpack.configs', [])):
        state = WebpackState(settings, root_package_name, name=extra_config)
        config.registry.webpack[extra_config] = state

    # Set up any static views
    for state in six.itervalues(config.registry.webpack):
        if state.static_view:
            config.add_static_view(name=state.static_view_name,
                                   path=state.static_view_path,
                                   cache_max_age=state.cache_max_age)

    config.add_request_method(get_webpack, 'webpack')
